server/src/main/java/com/linecorp/centraldogma/server/internal/httpapi/ContentServiceV1.java

```diff
@@ -0,0 +1,336 @@
+import com.linecorp.centraldogma.server.internal.command.Command;
+import com.linecorp.centraldogma.server.internal.command.CommandExecutor;
+import com.linecorp.centraldogma.server.internal.storage.project.ProjectManager;
+import com.linecorp.centraldogma.server.internal.storage.repository.ChangeConflictException;
+import com.linecorp.centraldogma.server.internal.storage.repository.FindOption;
+import com.linecorp.centraldogma.server.internal.storage.repository.Repository;
+
+public class ContentServiceV1 extends AbstractService {
+
+    public ContentServiceV1(ProjectManager projectManager, CommandExecutor executor) {
+        super(projectManager, executor);
+    }
+
+    /**
+     * GET /projects/{projectName}/repositories/{repository}/tree{path}?revision={revision}
+     *
+     * <p>Returns the list of files in the path.
+     */
+    @Get("regex:/projects/(?<projectName>[^/]+)/repos/(?<repoName>[^/]+)/tree(?<path>/.*$)")
+    public CompletionStage<List<EntryDto>> listFiles(@Param("projectName") String projectName,
+                                                     @Param("repoName") String repoName,
+                                                     @Param("path") String path,
+                                                     @Param("revision") @Default("-1") String revision) {
+        checkRepositoryExists(projectName, repoName);
+
+        final Repository repository = getRepository(projectName, repoName);
+        return listFiles(repository, path, new Revision(revision),
+                         ImmutableMap.of(FindOption.FETCH_CONTENT, false));
+    }
+
+    private CompletionStage<List<EntryDto>> listFiles(Repository repository, String path,
+                                                      Revision revision, Map<FindOption<?>, ?> options) {
+        final String pathPattern = path + (path.endsWith("/") ? '*' : "/*");
+        return repository.find(revision, pathPattern, options)
+                         .thenApply(
+                                 entries -> entries.values().stream()
+                                                   .filter(entry -> entry.type() != EntryType.DIRECTORY)
+                                                   .map(DtoConverter::convert).collect(toList()));
+    }
+
+    private void checkRepositoryExists(String projectName, String repoName) {
+        if (!projectManager().exists(projectName)) {
+            throw new HttpResponseException(HttpStatus.BAD_REQUEST);
+        }
+
+        if (!projectManager().get(projectName).repos().exists(repoName)) {
+            throw new HttpResponseException(HttpStatus.BAD_REQUEST);
+        }
+    }
+
+    /**
+     * PUT /projects/{projectName}/repos/{repoName}/contents{path}?revision={revision}
+     *
+     * <p>Adds or edit a file.
+     */
+    @Put("regex:/projects/(?<projectName>[^/]+)/repos/(?<repoName>[^/]+)/contents(?<path>/.*$)")
+    public CompletionStage<ContentDto> addOrEditFile(
```

Comment:
Because there's an input for revision, this is not idempotent. If I change this using `POST`,  it does not make sense to edit file. Do you have an idea for this?

Reply:
Or change the logic which returns just 200 OK when there's no change between files regardless of the revision?